state/protocol/protocol_state/kvstore.go

```diff
@@ -0,0 +1,108 @@
+package protocol_state
+
+import "github.com/onflow/flow-go/model/flow"
+
+// ViewBasedActivator allows setting value that will be active from specific view.
+type ViewBasedActivator[T any] struct {
+	Data           T
+	ActivationView uint64 // view at which data will be active
+}
+
+// VersionedEncodable defines the interface for a versioned key-value store independent
+// of the set of keys which are supported. This allows the storage layer to support
+// storing different key-value model versions within the same software version.
+type VersionedEncodable interface {
+	// VersionedEncode encodes the key-value store, returning the version separately
+	// from the encoded bytes.
+	// No errors are expected during normal operation.
+	VersionedEncode() (uint64, []byte, error)
+}
+
+// This file contains versioned read and read-write interfaces to the Protocol State's
+// key-value store and are used by the Protocol State Machine.
+//
+// When a key is added or removed, this requires a new protocol state version:
+//  - Create a new versioned model in models.go (eg. modelv3 if latest model is modelv2)
+//  - Update the Reader and API interfaces to include any new keys
+
+// Reader is the latest read-only interface to the Protocol State key-value store
+// at a particular block.
+// It is backward-compatible with all versioned model types defined in the models.go
+// for this software version.
+type Reader interface {
+	ID() flow.Identifier
+
+	// v0
+
+	VersionedEncodable
+
+	// GetProtocolStateVersion returns the current Protocol State version.
+	// The Protocol State version specifies the current version of the Protocol State,
+	// including the key-value store. Changes in the protocol state version
+	// correspond to changes in the set of key-value pairs which are supported,
+	// and which model is used for serialization.
+	// It can be updated by an UpdateKVStoreVersion service event.
+	GetProtocolStateVersion() uint64
+
+	// GetVersionUpgrade returns the upgrade version of protocol.
```

Comment:
Well, `GetVersionUpgrade`stores the activation view and version that can be used for next upgrade. It is NOT the version that was used to create underlying structure that implements API, for that we need to use `GetProtocolStateVersion`

Reply:
thanks for the explanation. I would suggest to update the documentation of `GetProtocolStateVersion` to 

```golang
// GetProtocolStateVersion returns the Protocol State Version that created the specific
// Snapshot backing this interface instance. Slightly simplified, the Protocol State
// Version defines the key-value store's data model (specifically, the set of all keys
// and the respective type for each corresponding value). 
// Generally, changes in the protocol state version correspond to changes in the set
// of key-value pairs which are supported, and which model is used for serialization.
// The protocol state version is updated by UpdateKVStoreVersion service events.
GetProtocolStateVersion() uint64
```

I have tried to avoid using "current" and instead talked about the  "Protocol State Version that created the specific
Snapshot backing this interface instance."